2018/2018_File_Generation_Features.md

```diff
@@ -0,0 +1,67 @@
+
+Edit test files are written to the directory indicated in `test_filepaths.yaml` based on edit type, such as /edits_files/test_files/Bank1/syntax/s300.txt. The file name indicates which edit the file is designed to test. If an edit has multiple conditions, a file will be made for each condition in the format /edits_files/test_files/Bank1/syntax/s301_1.txt.
+
+### Changing Edit Logic and the Classes Used for File Creation 
+The FileGenerator Class relies on four other classes, each with its own role in file creation. 
+
+-`lar_generator` in [`lar_generator.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/lar_generator.py) creates raw TS and LAR rows, the values of which are then manipulated to create clean test files or edit test files. 
+
+-`rules_engine` in [`rules_engine.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/rules_engine.py) contains logic for each edit. LAR and TS dataframes are passed into this class, and its methods output edit details that show among its fields the edit name, the status of the edit (failed or passed), the number of rows failed by the edit, if any, and the ULI's or 
+NULIs (loan ID) of the rows that fail the edit. Each method creates a results dataframe that holds the rows that have failed the specific edit. The class then wraps the results into an edit report that can be parsed by other methods in the repository.  
+
+- `lar_constraints` in [`lar_constraints.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/rules_engine.py) contains methods for creating clean LAR rows that conform to pass syntax or validity edits. New edit logic methods must be made in terms of changing a TS or LAR row. 
+
+- `test_data` in [`test_file_generator.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/test_file_generator.py) contains methods that manipulate LAR or TS dataframes to fail a specific syntax or validity edit. New edit logic methods must be made in terms of changing a TS or LAR dataframe.
+
+Each edit logic method in `lar_constraints`, `test_data`, and `rules_engine` is prefaced with 's', 'v', or 'q' to indicate a syntax, validity or quality edit respectively.
+
+### Validating Quality Edit Files and Creating Edit Reports
+The `test_data` logic for creating quality edit test files may cause rows to fail not only quality edits, but also syntax or validity edits in the creation process. One method in the FileGenerator class, `validate_quality_edit_file(quality_filename)` takes a quality edit test file from the quality filepaths directory in `test_filepaths.yaml`, drops rows that have syntax or validity edits, and duplicates the remaining clean rows to the length of the original file. The file is then saved in a new directory.
```

Comment:
After the first line:

The validate_quality_edit_file() method creates quality edit test files that do not have validity or syntax errors. This method uses the quality file path in `test_filepaths.yaml` to find a clean file for a specified quality edit. This clean file is used to source rows that pass validity and syntax edits that fail quality. This row is used as a base to construct a file of the length specified in [the appropriate config file](link). File outputs from this process are saved in the directory specified in [this config file](link).

Reply:
Sounds good - language has been added.